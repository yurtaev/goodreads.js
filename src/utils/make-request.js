// @flow


import { partial, anyPass, equals, not, isNil } from 'ramda'
const OAuth = require('oauth-1.0a')
const CryptoJS = require('crypto-js')

import { prepareResponse } from './prepare-response'

import type { ResponseAPI, MethodType } from './../../types/api'

// var oauth = OAuth({
//   consumer: {
//     key: 'consumer_key',
//     secret: 'consumer_secret',
//   },
//   signature_method: 'HMAC-SHA1',
//   hash_function: (base_string, key) => CryptoJS.HmacSHA1(base_string, key).toString(CryptoJS.enc.Base64),
// })

const getOAuthSigner = (credentials: Credentials) => OAuth({
  consumer: {
    ...credentials.consumer,
  },
  signature_method: 'HMAC-SHA1',
  hash_function: (base_string, key) => CryptoJS.HmacSHA1(base_string, key).toString(CryptoJS.enc.Base64),
})

// var token = {
//   key: 'user_access_key',
//   secret: 'user_access_secret',
// }

type Token = {|
  key: string,
  secret: string,
|}

/** Credentials */
export type Credentials = {
  consumer: {
    secret: string,
  },
  user: {
    key: string,
    secret: string,
  }
}

type Options = {|
  data?: any,
|}

const isPOST = equals('POST')
const isPUT = equals('PUT')
const isPATCH = equals('PATCH')
const isAllowBodyMethod = anyPass([isPOST, isPUT, isPATCH])

export const makeRequestOptions = (method: MethodType, url: string, credentials: Credentials, data: any = {}) => {
  const oauth = getOAuthSigner(credentials)
  const oauth_data = oauth.authorize({ url, method }, credentials.user)
  const oauthHeaders = oauth.toHeader(oauth_data)

  const options = {
    method,
    headers: { ...oauthHeaders },
  }

  if (isAllowBodyMethod(method) && not(isNil(data))) {
    const body = JSON.stringify(data)
    return {
      ...options,
      body,
    }
  }

  return options
}

export const makeAPIRequest = async (
  method: MethodType,
  url: string,
  credentials: Credentials,
  data?: any,
): Promise<ResponseAPI<any>> => {
  const options = makeRequestOptions(method, url, credentials, data)


  try {
    const response = await fetch(url, options)

    try {
      return await prepareResponse(response)
    } catch (err) {
      return Promise.reject(err)
    }
  } catch (err) {
    return Promise.reject({
      ok: false,
      status: -1,
      error: 'Network error',
      data: err,
    })
  }
}

export const makeGETRequest = partial(makeAPIRequest, ['GET'])
export const makePOSTRequest = partial(makeAPIRequest, ['POST'])
export const makePUTRequest = partial(makeAPIRequest, ['PUT'])
export const makePATCHRequest = partial(makeAPIRequest, ['PATCH'])
export const makeDELETERequest = partial(makeAPIRequest, ['DELETE'])
