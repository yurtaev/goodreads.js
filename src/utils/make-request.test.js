import MockDate from 'mockdate'

import { makeRequestOptions } from './make-request'

const data = {
  some: 'data',
  x: 123,
  y: false,
  z: null,
}

const token = 'test-token'
const url = 'https://example.com'

const credentials = {
  consumer: {
    secret: 'consumer_secret',
  },
  user: {
    key: 'user_key',
    secret: 'user_secret',
  }
}

describe('makeRequestOptions', () => {
  beforeAll(() => {
    MockDate.set('2017-01-31T00:00:00Z', -360) // 1/1/2017 Timezone +6

    const mockMath = Object.create(global.Math)
    mockMath.randomOld = Math.random
    mockMath.random = () => 0.5
    window.Math = mockMath
  })

  afterAll(() => {
    MockDate.reset()
    Math.random = Math.randomOld
  })

  test('GET', () => {
    expect(makeRequestOptions('GET', url, credentials, )).toMatchSnapshot('GET')
    expect(makeRequestOptions('GET', url, credentials)).toMatchSnapshot('GET token')
  })

  test('POST', () => {
    expect(makeRequestOptions('POST', url, credentials, data)).toMatchSnapshot('POST')
    expect(makeRequestOptions('POST', url, credentials, data)).toMatchSnapshot('POST token')
  })

  test('PUT', () => {
    expect(makeRequestOptions('PUT', url, credentials, data)).toMatchSnapshot('PUT')
    expect(makeRequestOptions('PUT', url, credentials, data)).toMatchSnapshot('PUT token')
  })

  test('PATCH', () => {
    expect(makeRequestOptions('PATCH', url, credentials, data)).toMatchSnapshot('PATCH')
    expect(makeRequestOptions('PATCH', url, credentials, data)).toMatchSnapshot('PATCH token')
  })

  test('DELETE', () => {
    expect(makeRequestOptions('DELETE', url, credentials, data)).toMatchSnapshot('DELETE')
    expect(makeRequestOptions('DELETE', url, credentials, data)).toMatchSnapshot('DELETE, token')
  })
})
