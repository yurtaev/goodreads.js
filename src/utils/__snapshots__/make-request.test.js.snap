// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DELETE 1`] = `
Object {
  "headers": Object {
    "Authorization": "OAuth oauth_consumer_key=\\"undefined\\", oauth_nonce=\\"FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\\", oauth_signature=\\"G5sOgo6LZKI0dgVxSA4a%2F0H5gmI%3D\\", oauth_signature_method=\\"HMAC-SHA1\\", oauth_timestamp=\\"1485820800\\", oauth_token=\\"user_key\\", oauth_version=\\"1.0\\"",
  },
  "method": "DELETE",
}
`;

exports[`DELETE, token 1`] = `
Object {
  "headers": Object {
    "Authorization": "OAuth oauth_consumer_key=\\"undefined\\", oauth_nonce=\\"FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\\", oauth_signature=\\"G5sOgo6LZKI0dgVxSA4a%2F0H5gmI%3D\\", oauth_signature_method=\\"HMAC-SHA1\\", oauth_timestamp=\\"1485820800\\", oauth_token=\\"user_key\\", oauth_version=\\"1.0\\"",
  },
  "method": "DELETE",
}
`;

exports[`GET 1`] = `
Object {
  "headers": Object {
    "Authorization": "OAuth oauth_consumer_key=\\"undefined\\", oauth_nonce=\\"FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\\", oauth_signature=\\"%2FJPxHSjmT5yiBl6WupGIKYJE4qQ%3D\\", oauth_signature_method=\\"HMAC-SHA1\\", oauth_timestamp=\\"1485820800\\", oauth_token=\\"user_key\\", oauth_version=\\"1.0\\"",
  },
  "method": "GET",
}
`;

exports[`GET token 1`] = `
Object {
  "headers": Object {
    "Authorization": "OAuth oauth_consumer_key=\\"undefined\\", oauth_nonce=\\"FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\\", oauth_signature=\\"%2FJPxHSjmT5yiBl6WupGIKYJE4qQ%3D\\", oauth_signature_method=\\"HMAC-SHA1\\", oauth_timestamp=\\"1485820800\\", oauth_token=\\"user_key\\", oauth_version=\\"1.0\\"",
  },
  "method": "GET",
}
`;

exports[`PATCH 1`] = `
Object {
  "body": "{\\"some\\":\\"data\\",\\"x\\":123,\\"y\\":false,\\"z\\":null}",
  "headers": Object {
    "Authorization": "OAuth oauth_consumer_key=\\"undefined\\", oauth_nonce=\\"FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\\", oauth_signature=\\"or6Amqgo%2BKDpn24f%2FbJN57Sdb4s%3D\\", oauth_signature_method=\\"HMAC-SHA1\\", oauth_timestamp=\\"1485820800\\", oauth_token=\\"user_key\\", oauth_version=\\"1.0\\"",
  },
  "method": "PATCH",
}
`;

exports[`PATCH token 1`] = `
Object {
  "body": "{\\"some\\":\\"data\\",\\"x\\":123,\\"y\\":false,\\"z\\":null}",
  "headers": Object {
    "Authorization": "OAuth oauth_consumer_key=\\"undefined\\", oauth_nonce=\\"FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\\", oauth_signature=\\"or6Amqgo%2BKDpn24f%2FbJN57Sdb4s%3D\\", oauth_signature_method=\\"HMAC-SHA1\\", oauth_timestamp=\\"1485820800\\", oauth_token=\\"user_key\\", oauth_version=\\"1.0\\"",
  },
  "method": "PATCH",
}
`;

exports[`POST 1`] = `
Object {
  "body": "{\\"some\\":\\"data\\",\\"x\\":123,\\"y\\":false,\\"z\\":null}",
  "headers": Object {
    "Authorization": "OAuth oauth_consumer_key=\\"undefined\\", oauth_nonce=\\"FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\\", oauth_signature=\\"jbom%2FqUYh4CrabBSl7Nzq4bcOMg%3D\\", oauth_signature_method=\\"HMAC-SHA1\\", oauth_timestamp=\\"1485820800\\", oauth_token=\\"user_key\\", oauth_version=\\"1.0\\"",
  },
  "method": "POST",
}
`;

exports[`POST token 1`] = `
Object {
  "body": "{\\"some\\":\\"data\\",\\"x\\":123,\\"y\\":false,\\"z\\":null}",
  "headers": Object {
    "Authorization": "OAuth oauth_consumer_key=\\"undefined\\", oauth_nonce=\\"FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\\", oauth_signature=\\"jbom%2FqUYh4CrabBSl7Nzq4bcOMg%3D\\", oauth_signature_method=\\"HMAC-SHA1\\", oauth_timestamp=\\"1485820800\\", oauth_token=\\"user_key\\", oauth_version=\\"1.0\\"",
  },
  "method": "POST",
}
`;

exports[`PUT 1`] = `
Object {
  "body": "{\\"some\\":\\"data\\",\\"x\\":123,\\"y\\":false,\\"z\\":null}",
  "headers": Object {
    "Authorization": "OAuth oauth_consumer_key=\\"undefined\\", oauth_nonce=\\"FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\\", oauth_signature=\\"8Ip2MZTobJZHSezx5kaV8HLgYXc%3D\\", oauth_signature_method=\\"HMAC-SHA1\\", oauth_timestamp=\\"1485820800\\", oauth_token=\\"user_key\\", oauth_version=\\"1.0\\"",
  },
  "method": "PUT",
}
`;

exports[`PUT token 1`] = `
Object {
  "body": "{\\"some\\":\\"data\\",\\"x\\":123,\\"y\\":false,\\"z\\":null}",
  "headers": Object {
    "Authorization": "OAuth oauth_consumer_key=\\"undefined\\", oauth_nonce=\\"FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\\", oauth_signature=\\"8Ip2MZTobJZHSezx5kaV8HLgYXc%3D\\", oauth_signature_method=\\"HMAC-SHA1\\", oauth_timestamp=\\"1485820800\\", oauth_token=\\"user_key\\", oauth_version=\\"1.0\\"",
  },
  "method": "PUT",
}
`;
